apps/backend/app/services/s3_service.py

```
@@ -44,16 +48,16 @@ def download_file_from_s3(s3_url: str) -> str:
-        s3_pattern = r"https://([^\.]+)\.s3\.amazonaws\.com/([^?]+)"
-        match = re.match(s3_pattern, s3_url)
-
-        if not match:
+        # SECURITY: Validate S3 URL format and bucket whitelist.
+        # parse_s3_url handles amazonaws and endpoint-style (MinIO/LocalStack)
+        # URLs with exact-origin endpoint matching; URLs it cannot attribute
+        # to a bucket are rejected here.
+        try:
+            bucket_name, object_key = parse_s3_url(s3_url)
+        except ValueError:
+            raise ValueError(f"Invalid S3 URL format: {s3_url}")
```

Ruff (0.15.15)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)